recipe-client-addon/test/browser/browser_NormandyApi.js

```diff
@@ -3,19 +3,13 @@
 const {utils: Cu} = Components;
 Cu.import("resource://shield-recipe-client/lib/NormandyApi.jsm", this);
 
-add_task(function* () {
+add_task(function* test_get() {
   // Point the add-on to the test server.
-  yield SpecialPowers.pushPrefEnv({
-    set: [
-      [
-        "extensions.shield-recipe-client.api_url",
-        "http://mochi.test:8888/browser/browser/extensions/shield-recipe-client/test",
-      ],
-    ],
-  });
+  const url = "http://mochi.test:8888/browser/browser/extensions/shield-recipe-client/test";
+  yield SpecialPowers.pushPrefEnv({set: [["extensions.shield-recipe-client.api_url", url]]});
 
   // Test that NormandyApi can fetch from the test server.
-  const response = yield NormandyApi.get("browser/test_server.sjs");
+  const response = yield NormandyApi.get(`${url}/browser/test_server.sjs`);
   const data = yield response.json();
-  Assert.deepEqual(data, {test: "data"}, "NormandyApi returned incorrect server data.");
+  Assert.deepEqual(data, {test: "data"}, "NormandyApi returned correct server data.");
```

Comment:
You can use an sjs file ( see https://developer.mozilla.org/en-US/docs/Mozilla/httpd.js/HTTP_server_for_unit_tests ) to generate dummy responses for use in a test. It's true that they could get out of date with the server, but presumably when you're updating the server you update the client, and if you update the client without updating the mock server, you'd break things.

It's true that you would only ever guarantee that the test works with the mock server you'd write and so you wouldn't catch subtle differences between the mock server and the 'real' one, but you'd be able to have a bit more confidence in:
1) client changes that should continue to work with the same server
2) server changes that update the mock server in the 'same way' - you get infra to guarantee things work with your mock server, and so now in review you check the mock server is a reasonable representation of the real server, which is easier to do as a human being than "does my client code interact with this orthogonal server code in the right way".

I hope that reasoning makes some sense, though I also think it's something about which reasonable people can disagree. :-)